### PR TITLE
Experimental fit tutorial

### PR DESCRIPTION
**Pull request by abocquet**

### CODE REVIEW

**Review by pierreguilmin — COMMENTED** [review]
Nice!! I like the figures and the phrasing. I think this still needs a bit more work to be as striking as it could be, but it's a super good start. A few high-level comments before doing a detailed review later (probably next week):
- Can we fit more parameters using these curves? Our library shines at estimating many parameters, there's not much gain at doing gradient descent to estimate a single parameter.
- The optimisation time is very long for such a small system, we should probably JIT the optimisation loop?
- Simulated and synthetic data should look roughly the same (same number of curves).
- I think this tutorial is great, however IMO we should not think about as "a tutorial on how to perform gradient descent with dynamiqs". To me, this is an "example application" tutorial: how dynamiqs helped in a real life situation, on real data. We can make a section "Example application" rather than "Advanced". Advanced would be things like: "how to do robust quantum optimal control", "how to fit parameters with a second-order method by computing the Hessian", etc... which explain a method based on very simple systems. And a tutorial to learn how to perform gradient descent with dynamiqs should be in "Basics", it needs (1) a very simple system (the dissipative cat qubit is very niche), (2) fancier gradient descent, (3) random initial seeds. That being said, it's great to have example application tutorials, it's usually the ones that are the most convincing for experienced users. See e.g. https://docs.q-ctrl.com/boulder-opal/user-guides vs https://docs.q-ctrl.com/boulder-opal/application-notes.
- The library name is `dynamiqs` and not `Dynamiqs`, I think we should not use the capitalized version.

File: docs/tutorials/experimental-fits.md [hunk 1]
```diff
@@ -0,0 +1,436 @@
+# Experimental data fits
+
+Dynamiqs enables rapid simulations and facilitates automatic gradient computation, making it highly suitable for data fitting tasks. This tutorial will guide you through the process of simulating and fitting real-world experimental data, step by step.
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:3]:
```suggestion
dynamiqs enables fast simulations of quantum systems with automatic gradient computation, making it highly suitable for data fitting tasks. This tutorial will guide you through the process of simulating and fitting real-world experimental data, step by step.
```

File: docs/tutorials/experimental-fits.md [hunk 2]
```diff
@@ -0,0 +1,436 @@
+# Experimental data fits
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:1]:
```suggestion
# Fitting a cat qubit experiment with real-world data
```

File: mkdocs.yml [hunk 3]
```diff
@@ -121,6 +121,8 @@ nav:
           - tutorials/open-system.md
           - tutorials/batching-simulations.md
           - tutorials/computing-gradients.md
+      - Advanced:
+          - tutorials/experimental-fits.md
```
Comment (gautierronan) [inline mkdocs.yml:125]:
```suggestion
      - Example applications:
          - applications/experimental-fits.md
```

File: docs/tutorials/experimental-fits.md [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+
+Dynamiqs enables rapid simulations and facilitates automatic gradient computation, making it highly suitable for data fitting tasks. This tutorial will guide you through the process of simulating and fitting real-world experimental data, step by step.
+
+## Context of the experiment
+
+!!! note Info
+    This part introduces the kind of data and process that we will fit in the notebook. If you are only interested in the code you can skip it.
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:8]:
```suggestion
    This section introduces the kind of data and process that we will fit in the notebook. If you are only interested in the code you can skip the next section.
```

File: docs/tutorials/experimental-fits.md [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+
+!!! example "Diving deep 🤿"
+    You can discover further details about this experiment in the original research paper from which this data was sourced [^1]
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:11]: I think this is a bit heavy/insistent: you should definitely cite the paper but probably within the next paragraph.

File: docs/tutorials/experimental-fits.md [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+
+The data we will work with are deflation curves. Our system is composed of a high lifetime cavity - the memory -  coupled to a lossy one - the buffer - in vacuum state. We denote $a$ (respectively $b$) the memory (respectively buffer) annihilation operator. The memory and the buffer are coupled through a two to one photon exchange mechanism $g_2 a^2 b^\dagger + h.c.$. The buffer is a highly-dissipative mode, that we model with the dissipator $D[\sqrt{\kappa_b} b]$. Both cavities have a finite temperature, expressed in thermal population.
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:13]: Start with a short but higher-view description of the experiment (in which you can cite the paper). This is a bit rough to start with "The data we will work with are deflation curves" which doesn't mean anything to the regular physicist.

File: docs/tutorials/experimental-fits.md [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+
+At the beginning of the experiment the memory is populated with a coherent state and the buffer is in thermal vacuum. We enable for a period of two photon exchange mechanism. During this period, pairs of photons leave the memory and are converted into a single buffer photon, that is then dissipated into the environment. The memory photon number $a^\dagger a$ is measured at different times and saved into the `data` variable. Our goal is to infer $g_2$ from the data.
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:15]: "We enable for a period of two photon exchange mechanism"

Not sure what this means. Do you mean you turn on the two-photon exchange mechanism?

File: docs/tutorials/experimental-fits.md [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+
+![Pulse sequence](/figs-docs/g2-pulse-sequence.png)
+_Pulse sequence of the experiment, taken from [^1]_
+
+[^1]:  [Quantum control of a cat-qubit with bit-flip times exceeding ten seconds, arxiv:2307.06617](https://arxiv.org/pdf/2307.06617.pdf)
+
+## Step by step solution
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:22]:
```suggestion
## Fitting real-world data, step by step
```

File: docs/tutorials/experimental-fits.md [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:23]: Can you add a short introduction to what you will be doing in this section?

File: docs/tutorials/experimental-fits.md [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+```python
+thermal_a, thermal_b = 0.1, 0.011
+kappa_a = 9.3e-3 * kHz
+kappa_b = 2.6 * MHz
+```
+
+Instanciate our bosonic annihilation operators:
+
+```python
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:54]: If you only have single-line comments about each of these things, I would suggest putting them all in a single code block, with python comments above. It's easier to read and to follow.

File: docs/tutorials/experimental-fits.md [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+Na, Nb = 15, 7
+a, b = dq.destroy(Na, Nb))
+```
+
+### Loading the data
+
+We provide real experimental data but also leave the possibility to simply simulate them, you can choose the one that suits you best
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:61]:
```suggestion
We now initialize the experimental data with which we will work. There are two possible options to do so: either load the actual experimental data from [^1]_, or generate synthetic data from a noisy master equation simulation.
```

File: docs/tutorials/experimental-fits.md [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+
+#### Option 1: Real world data
+
+Before running the code, you can download the data at this link (TODO: add link)
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:65]: I guess the data should be put somewhere on the github (probably in the same folder), so people don't have to open a link to some scary website.

File: docs/tutorials/experimental-fits.md [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+
+Load the data
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:67]: Same comment as before: either say a bit more about what you're doing, or put these comments inside the code-block. Right now it's a bit rough around the edges.

File: docs/tutorials/experimental-fits.md [hunk 14]
```diff
@@ -0,0 +1,436 @@
...
+
+plt.plot(time, data, "+")
+plt.grid()
+plt.xlabel("Time [us]")
+plt.ylabel("Photon number")
+renderfig('g2-real-data')
+```
+
+% invisible-code-block: python
+% exp_time, exp_data, exp_alpha2 = time, data, alpha2
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:93]: Since this doesn't work yet, either we first fix it in a separate PR, or we remove this altogether (for now).

File: docs/tutorials/experimental-fits.md [hunk 15]
```diff
@@ -0,0 +1,436 @@
...
+
+![Loaded data](/figs-docs/g2-real-data.png)
+
+#### Option 2: Synthetic data
+
+Create the synthetic data 
+
+```python
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:101]: This whole block/section is lacking comments or descriptions.

File: docs/tutorials/experimental-fits.md [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+time_np = time
+for i in range(len(alpha2)):
+    plt.plot(time_np, data[:, i].real, f"+C{i}")
+    plt.plot(time_np, result.expects[i, 0].real, f"C{i}")
+plt.grid()
+plt.xlabel("Time [us]")
+plt.ylabel("Photon number")
+renderfig('g2-initial-guess')
+```
+![g2 initial guess](/figs-docs/g2-initial-guess.png)
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:186]: This figure looks a bit silly IMO. Maybe taking $g_2$ to be much larger (say 5 MHz) would be better?

File: docs/tutorials/experimental-fits.md [hunk 17]
```diff
@@ -0,0 +1,436 @@
...
+
+
+% skip: start
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:189]: Same comment as before: either we fix it or we remove it.

File: docs/tutorials/experimental-fits.md [hunk 18]
```diff
@@ -0,0 +1,436 @@
...
+costs = np.array(costs)
+g2s = np.array(g2s)
+```
+we use a plain gradient descent algorithm for simplicity but one could imagine using any other algorithm, such as the ones provided in [Optax](https://github.com/google-deepmind/optax).
+
+!!! note Optimisation time
+    The full optimisation runs in less than one minute on a NVIDIA GeForce RTX 4090 and around 5 minutes on a M1 Mac.
+
+
+We can plot the loss to ensure we converged correctly
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:217]: Where is this plot?

File: docs/tutorials/experimental-fits.md [hunk 19]
```diff
@@ -0,0 +1,436 @@
...
+print(g2 / MHz)
+```
+
+!!! example "Final note"
+    $g_2$ value differs from the one reported in
+    [arxiv:2307.06617](https://arxiv.org/pdf/2307.06617.pdf)
+    by approximately 15%. This discrepancy is attributed to the unavailability 
+    of Dynamiqs for performing fits at the time of writing the article. 
+    Consequently, the authors were unable to employ gradient descent for their
+    fit and had to resort to alternative methods.
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:257]: I would remove this altogether. You're just giving the baton pour te faire battre here.

File: docs/tutorials/experimental-fits.md [hunk 20]
```diff
@@ -0,0 +1,436 @@
...
+
+
+## Full code
```
Comment (gautierronan) [inline docs/tutorials/experimental-fits.md:260]: Ideally we'll have a link at some point for this, but for now it's fine I think.

**Review by gautierronan — CHANGES_REQUESTED** [review]
I overall agree with @pierreguilmin: it's a really nice tutorial to have. It really shows a new kind of application that is enabled with dynamiqs. So I'd be down to merge this before the MM actually.

However, the tutorial in itself is a bit rough around the edges for now. It's lacking comments and details: you're not explaining much and it makes it hard to follow for the reader (think of a master student working on e.g. transmons). I provided a few suggestions here and there, but it's just overall guidelines.

Regarding @pierreguilmin's comments, could you indeed also fit the $\kappa_b$ as well for instance? Or even $\kappa_a$ and $\kappa_b$? I agree with his other comments otherwise.

File: docs/tutorials/experimental-fits.md (hunk 7 above)
Comment (abocquet) [reply docs/tutorials/experimental-fits.md:15]: Changed to 
> At the beginning of the experiment the memory is populated with a coherent state and the buffer is in thermal vacuum. We activate the two photons exchange mechanism. Pairs of photons leave the memory and are converted into a single buffer photon, that is then dissipated into the environment. The memory photon number $a^\dagger a$ is measured at different times and saved into the `data` variable. Our goal is to infer $g_2$ from the data.

File: docs/tutorials/experimental-fits.md (hunk 12 above)
Comment (abocquet) [reply docs/tutorials/experimental-fits.md:65]: Indeed, but the files are nowhere on github yet

File: docs/tutorials/experimental-fits.md (hunk 14 above)
Comment (abocquet) [reply docs/tutorials/experimental-fits.md:93]: The docs won't run without this

File: docs/tutorials/experimental-fits.md (hunk 6 above)
Comment (abocquet) [reply docs/tutorials/experimental-fits.md:13]: >Our system is composed of a high lifetime cavity - the memory -  coupled to a lossy one - the buffer - in vacuum state. We denote $a$ (respectively $b$) the memory (respectively buffer) annihilation operator. The memory and the buffer are coupled through a two to one photon exchange mechanism $g_2 a^2 b^\dagger + h.c.$. The buffer is a highly-dissipative mode, that we model with the dissipator $D[\sqrt{\kappa_b} b]$. Both cavities have a finite temperature, expressed in thermal population. For more details, please see [^1].

> At the beginning of the experiment the memory is populated with a coherent state and the buffer is in thermal vacuum. We activate the two photons exchange mechanism. Pairs of photons leave the memory and are converted into a single buffer photon, that is then dissipated into the environment. The memory photon number $a^\dagger a$ is measured at different times and saved into the `data` variable. Our goal is to infer $g_2$ from the data.

File: docs/tutorials/experimental-fits.md (hunk 9 above)
Comment (abocquet) [reply docs/tutorials/experimental-fits.md:23]: > In this notebook, we will load the experimental data (or simulate some), encode the two photon exchange model in dynamiqs, simulate the system evolution and measure the number of photons at different instants. We will then use gradient descent to fit the $g_2$ parameter to the data.

File: docs/tutorials/experimental-fits.md (hunk 14 above)
Comment (abocquet) [reply docs/tutorials/experimental-fits.md:93]: Yes let's fix it 👍

File: docs/tutorials/experimental-fits.md (hunk 16 above)
Comment (abocquet) [reply docs/tutorials/experimental-fits.md:186]: It makes sense to show that the fit is bad in a first place. At 5MHz you have a batch of oscillations, it feels like a deliberately bad choice

File: docs/tutorials/experimental-fits.md (hunk 18 above)
Comment (abocquet) [reply docs/tutorials/experimental-fits.md:217]: Fixed

**Comment by abocquet** [comment]
Thank you for your feedback. Most of it should be addressed by now. I don't know if I'm down adding other parameters fit because that is not how we did it, and experimentally we like fit parameters one by one instead of all at once. I need to think about it

File: docs/tutorials/experimental-fits.md (hunk 16 above)
Comment (gautierronan) [reply docs/tutorials/experimental-fits.md]: I think it looks better, we clearly see it's not the right value no?

Comment (abocquet) [reply docs/tutorials/experimental-fits.md]: We also see it with the low value, and it's more complicated  to me. It makes more sense to start from lower values because it entails less non linearities in you system so it's easier to understand

**Comment by gautierronan** [comment]
@abocquet I made several corrections in 6376fabb6e9884984194c8f675e1088e0fa528ff. I'll let you look at the result, and check that you agree with the changes I made. Feel free to revert any change you don't agree with.

A few things left to do before merging:

- [ ] Add experimental data on the github, and add the link to the tutorial
- [ ] Update all figures and final g2 value with the changes I made (I didn't have access to the experimental data)
- [ ] Fix the `% skip: start` and `% skip: end` (can be done in another PR, but ideally before merging this).
- [ ] Make a final check that the provided code runs smoothly, without bugs.

Thanks for your work on this, I think it's going to be a great tutorial !

File: docs/applications/experimental-fits.md [hunk 21]
```diff
@@ -128,122 +149,123 @@ renderfig('g2-synthetic-data')
...
 plt.ylabel("Photon number")
-
-print(g2 / MHz)
+```
+```text
+g2 = 0.785 MHz
 ```
 ![Fit result](/figs-docs/g2-fit-result.png)
 
 !!! example "Final note"
```
Comment (abocquet) [inline docs/applications/experimental-fits.md:264]:
```suggestion

```

**Review by abocquet — COMMENTED** [review]
I cannot approve but I do

File: docs/applications/experimental-fits.md [hunk 22]
```diff
@@ -128,122 +149,123 @@ renderfig('g2-synthetic-data')
...
-g2s = np.array(g2s)
+# Save the final cost and convert to jnp arrays
+costs.append(cost_fn(g2, data))
+costs, g2s = jnp.array(costs), jnp.array(g2s)
 ```
-we use a plain gradient descent algorithm for simplicity but one could imagine using any other algorithm, such as the ones provided in [Optax](https://github.com/google-deepmind/optax).
 
 !!! note Optimisation time
-    The full optimisation runs in less than one minute on a NVIDIA GeForce RTX 4090 and around 5 minutes on a M1 Mac.
+    The full optimisation runs in about 20 seconds on a NVIDIA GeForce RTX 4090 and around 15 minutes on a M2 Macbook air.
```
[the remark this answers is not in the document]
Comment (gautierronan) [reply docs/applications/experimental-fits.md:224]: No idea ! I thought the M2 were faster ;p Maybe something to do with my JAX install (e.g. multi-threading or not). We can keep the previous values if you prefer.

[the remark this answers is not in the document]
Comment (abocquet) [reply docs/applications/experimental-fits.md:224]: Let's take the best available numbers 🚀
